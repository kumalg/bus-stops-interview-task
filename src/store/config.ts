import type { BusStop, BusStopsFetchStatus } from '@/types';

export enum StoreMutation {
    SetBusStops = 'SET_BUS_STOPS',
    SetStatus = 'SET_STATUS',
    SetLineStops = 'SET_LINE_STOPS',
    SetLineStopTimes = 'SET_LINE_STOP_TIMES'
}

export enum StoreAction {
    FetchBusStops = 'fetchBusStops',
    GetLineStops = 'getLineStops',
    GetTimesForLineStop = 'getTimesForLineStop'
}

export type LineStopsKey = BusStop['line'];
export type LineStopsValue = BusStop['stop'][];

export type LineStopTimesKey = {
    line: BusStop['line'];
    stop: BusStop['stop'];
};
export type LineStopTimesValue = BusStop['time'][];

export type StoreState = {
    busStops: BusStop[];
    status: BusStopsFetchStatus;
    lineStops: {
        [key in LineStopsKey]: LineStopsValue;
    };
    lineStopTimes: Map<LineStopTimesKey, LineStopTimesValue>;
};
