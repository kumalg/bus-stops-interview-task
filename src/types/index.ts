export type BusStop = {
  line: number,
  stop: string,
  order: number,
  time: string
}

export type Order = 'asc' | 'desc'