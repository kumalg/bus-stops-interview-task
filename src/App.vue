<template>
    <div class="app">
        <Transition name="fade" mode="out-in">
            <Dashboard v-if="status === BusStopsFetchStatus.FETCHED" class="dashboard" />
            <div v-else-if="status === BusStopsFetchStatus.FETCHING" class="message-loading">
                Loading data
            </div>
            <div v-else-if="status === BusStopsFetchStatus.ERROR" class="message-error">
                Error occurred while loading data
                <BaseButton @click="fetchData()">Fetch again</BaseButton>
            </div>
        </Transition>
    </div>
</template>

<script setup lang="ts">
import { computed, onBeforeMount } from 'vue';

import { StoreAction } from '@/store/config';
import { useStore } from '@/store';

import { BusStopsFetchStatus } from '@/types';

import BaseButton from '@/components/BaseButton/BaseButton.vue';
import Dashboard from './components/Dashboard/Dashboard.vue';

const store = useStore();
const status = computed(() => store.state.status);

const fetchData = () => {
    store.dispatch(StoreAction.FetchBusStops);
};

onBeforeMount(() => {
    fetchData();
});
</script>

<style lang="scss" scoped>
.app {
    padding: 2.5rem 2rem;

    .message-loading,
    .message-error {
        display: flex;
        flex-direction: column;
        align-items: center;
        font-size: 1rem;
        gap: 1rem;
    }

    @include desktop {
        height: 100vh;
        min-height: 640px;
    }
}
</style>
