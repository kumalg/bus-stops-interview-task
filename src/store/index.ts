import type { BusStop } from '@/types';
import type { StoreState } from '@/store/config';

import { createStore, useStore as baseUseStore, Store } from 'vuex';
import { InjectionKey } from 'vue';

import { toSorted, toSortedTimes } from '@/assets/sort';
import { validateBusStops } from '@/assets/validator';

import { StoreAction, StoreMutation } from '@/store/config';

import { BusStopsFetchStatus } from '@/types';

export const key: InjectionKey<Store<StoreState>> = Symbol();

export const store = createStore<StoreState>({
    state: {
        busStops: [],
        status: BusStopsFetchStatus.UNFETCHED,
        lineStops: {},
        lineStopTimes: new Map()
    },
    getters: {
        lines: (state) => toSorted(new Set(state.busStops.map(({ line }) => line))),
        stops: (state) => toSorted(new Set(state.busStops.map(({ stop }) => stop)))
    },
    mutations: {
        [StoreMutation.SetBusStops](state, value: BusStop[]) {
            state.busStops = value;
        },
        [StoreMutation.SetStatus](state, value: BusStopsFetchStatus) {
            state.status = value;
        },
        [StoreMutation.SetLineStops](
            state,
            { line, stops }: { line: BusStop['line']; stops: BusStop['stop'][] }
        ) {
            state.lineStops[line] = stops;
        },
        [StoreMutation.SetLineStopTimes](
            state,
            {
                key,
                times
            }: { key: { line: BusStop['line']; stop: BusStop['stop'] }; times: BusStop['time'][] }
        ) {
            state.lineStopTimes.set(key, times);
        }
    },
    actions: {
        async [StoreAction.FetchBusStops]({ state, commit }) {
            if (state.status === BusStopsFetchStatus.FETCHING) {
                return;
            }

            commit(StoreMutation.SetStatus, BusStopsFetchStatus.FETCHING);

            try {
                const response = await fetch('http://localhost:3000/stops');

                if (!response.ok) {
                    commit(StoreMutation.SetStatus, BusStopsFetchStatus.ERROR);
                }

                const json = await response.json();

                if (!validateBusStops(json)) {
                    commit(StoreMutation.SetStatus, BusStopsFetchStatus.ERROR);
                } else {
                    commit(StoreMutation.SetBusStops, json);
                    commit(StoreMutation.SetStatus, BusStopsFetchStatus.FETCHED);
                }
            } catch {
                commit(StoreMutation.SetStatus, BusStopsFetchStatus.ERROR);
            }
        },
        [StoreAction.GetLineStops]({ state, commit }, line: number) {
            if (line in state.lineStops && typeof state.lineStops[line] === 'object') {
                return state.lineStops[line];
            }

            const stopFullObjects = state.busStops.filter((busStop) => busStop.line === line);
            stopFullObjects.sort(({ order: aOrder }, { order: bOrder }) => aOrder - bOrder);

            const stops = [...new Set(stopFullObjects.map(({ stop }) => stop))];

            commit(StoreMutation.SetLineStops, { line, stops });

            return stops;
        },
        [StoreAction.GetTimesForLineStop](
            { state, commit },
            key: { line: BusStop['line']; stop: BusStop['stop'] }
        ) {
            if (state.lineStopTimes.has(key)) {
                return state.lineStopTimes.get(key);
            }

            let times;
            try {
                times = toSortedTimes(
                    new Set(
                        state.busStops
                            .filter(({ line, stop }) => line === key.line && stop === key.stop)
                            .map(({ time }) => time)
                    )
                );
            } catch {
                return [];
            }

            commit(StoreMutation.SetLineStopTimes, { key, times });

            return times;
        }
    },
    modules: {}
});

export const useStore = () => {
    return baseUseStore(key);
};
