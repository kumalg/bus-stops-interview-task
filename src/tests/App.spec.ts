import { shallowMount } from '@vue/test-utils';
import { createStore } from 'vuex';

import { BusStopsFetchStatus } from '@/types';

import { StoreAction, StoreState } from '@/store/config';
import { key } from '@/store';

import App from '@/App.vue';

describe('App', () => {
    const fetchBusStopsMock = jest.fn();

    const AppFactory = ({
        status = BusStopsFetchStatus.UNFETCHED
    }: {
        status?: BusStopsFetchStatus;
    } = {}) => {
        const state: StoreState = {
            busStops: [],
            status,
            lineStops: new Map(),
            lineStopTimes: new Map()
        };

        const store = createStore({
            state,
            actions: {
                [StoreAction.FetchBusStops]: fetchBusStopsMock
            }
        });

        return shallowMount(App, {
            global: {
                provide: {
                    [key as unknown as string]: store
                }
            }
        });
    };

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('renders proper html markup', () => {
        const wrapper = AppFactory({
            status: BusStopsFetchStatus.FETCHED
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it('fetch data function has been called', () => {
        AppFactory();

        expect(fetchBusStopsMock).toHaveBeenCalledTimes(1);
    });

    it('should render app container if status is fetched', () => {
        const wrapper = AppFactory({
            status: BusStopsFetchStatus.FETCHED
        });

        expect(wrapper.find('.dashboard').exists()).toBe(true);
        expect(wrapper.find('.message-loading').exists()).toBe(false);
        expect(wrapper.find('.message-error').exists()).toBe(false);
    });

    it('should render loading message if status is fetching', () => {
        const wrapper = AppFactory({
            status: BusStopsFetchStatus.FETCHING
        });

        expect(wrapper.find('.dashboard').exists()).toBe(false);
        expect(wrapper.find('.message-loading').exists()).toBe(true);
        expect(wrapper.find('.message-error').exists()).toBe(false);
    });

    it('should render error message if status is error', () => {
        const wrapper = AppFactory({
            status: BusStopsFetchStatus.ERROR
        });

        expect(wrapper.find('.dashboard').exists()).toBe(false);
        expect(wrapper.find('.message-loading').exists()).toBe(false);
        expect(wrapper.find('.message-error').exists()).toBe(true);
    });

    it('should render nothing if status is unfetched', () => {
        const wrapper = AppFactory({
            status: BusStopsFetchStatus.UNFETCHED
        });

        expect(wrapper.find('.dashboard').exists()).toBe(false);
        expect(wrapper.find('.message-loading').exists()).toBe(false);
        expect(wrapper.find('.message-error').exists()).toBe(false);
    });
});
