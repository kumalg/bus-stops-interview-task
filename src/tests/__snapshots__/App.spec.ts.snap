// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders proper html markup 1`] = `
<div
  class="app"
>
  <transition-stub
    appear="false"
    css="true"
    mode="out-in"
    name="fade"
    persisted="false"
  >
    <dashboard-stub
      class="dashboard"
    />
  </transition-stub>
</div>
`;
